dask/array/reductions.py

```diff
@@ -31,30 +34,73 @@ def reduction(x, chunk, aggregate, axis=None, keepdims=None, dtype=None):
     if dtype and 'dtype' in getargspec(aggregate).args:
         aggregate = partial(aggregate, dtype=dtype)
 
-    chunk2 = partial(chunk, axis=axis, keepdims=True)
-    aggregate2 = partial(aggregate, axis=axis, keepdims=keepdims)
+    # Normalize axes
+    if max_leaves is None:
+        max_leaves = dict(zip(range(x.ndim), x.numblocks))
+    if isinstance(max_leaves, int):
+        n = builtins.max(int(len(axis) ** (1/max_leaves)), 2)
+        max_leaves = dict.fromkeys(axis, n)
+    max_leaves = dict((k, v) for (k, v) in max_leaves.items() if k in axis)
 
+    # Map chunk across all blocks
     inds = tuple(range(x.ndim))
-    tmp = atop(chunk2, inds, x, inds)
-
-    inds2 = tuple(i for i in inds if i not in axis)
-
-    result = atop(compose(aggregate2, partial(_concatenate2, axes=axis)),
-                  inds2, tmp, inds, dtype=dtype)
-
-    if keepdims:
-        dsk = result.dask.copy()
-        for k in flatten(result._keys()):
-            k2 = (k[0],) + insert_many(k[1:], axis, 0)
-            dsk[k2] = dsk.pop(k)
-        chunks = insert_many(result.chunks, axis, [1])
-        return Array(dsk, result.name, chunks=chunks, dtype=dtype)
-    else:
-        return result
+    tmp = atop(partial(chunk, axis=axis, keepdims=True), inds, x, inds)
```

Comment:
If you feel particularly underburdened we could move the `partial` call here into `atop`, allowing it to support keyword args.